pilot/pkg/serviceregistry/kube/controller.go

```diff
@@ -665,15 +759,74 @@ func (c *Controller) AppendInstanceHandler(f func(*model.ServiceInstance, model.
 			return nil
 		}
 
-		log.Infof("Handle endpoint %s in namespace %s -> %v", ep.Name, ep.Namespace, ep.Subsets)
-		if item, exists := c.serviceByKey(ep.Name, ep.Namespace); exists {
-			if svc := convertService(*item, c.domainSuffix); svc != nil {
-				// TODO: we're passing an incomplete instance to the
-				// handler since endpoints is an aggregate structure
-				f(&model.ServiceInstance{Service: svc}, event)
+		if directEDS {
```

Comment:
Previously (in else branch), we don't have any Envoy XDS specific logic. But now we just inline the Envoy business logic here, the service registry controller now have some Envoy code. 

Can we instead use a pubsub model to make them separate? i.e. no `XDSUpdater` here, no need to call them, only need to deliver the updates.

Reply:
I don't think it's a good idea: I don't expect this code to be a general library to be used with  arbitrary things, and it's not worth the cost. I think that was a major cause of the performance issues we see -the attempt to abstract things without taking into account the needs of pilot. 

The registry exist to feed info into envoy via XDS in the most efficient way possible.